### PR TITLE
feat: add integration with Redux DevTools

### PR DESCRIPTION
**Pull request by pac-guerreiro**
### Details
This PR adds the ability to use React Native Onyx with Redux DevTools to debug storage changes.

Features:

- View `SET`, `MULTISET`, `MERGE`, `REMOVE` and `CLEAR` actions in DevTools: payload data corresponds to the data which was written into the storage;

- View state diffs when an action is selected;

- View current state at the time an action was called;

- Time travelling is not possible (yet)

### Related Issues
None

### Automated Tests
Not needed because this doesn't affect internal behaviour

### Linked PRs
None

### CODE REVIEW

**Comment by pac-guerreiro** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by AndrewGable** [comment]
Can you please add some documentation on how to get this setup as a developer?

**Comment by pac-guerreiro** [comment]
@AndrewGable 

I just added some instructions to the `README.md` and also updated the `API.md`.

I decided to differentiate between debugging setState and using Redux DevTools extension.

Let me know what you think! 😄

**Comment by PauloGasparSv** [comment]
Hey @pac-guerreiro we have a conflict now, can you please take a look?

I'll review this tomorrow, got back from OOO today and had to prioritize other issues/reviews.

File: README.md [hunk 1]
```diff
@@ -321,8 +321,18 @@ Sample output of `Onyx.printMetrics()`
...
+
 It can be useful to log why Onyx is calling `setState()` on a particular React component so that we can understand which key changed, what changed about the value, and the connected component that ultimately rendered as a result. When used correctly this can help isolate problem areas and unnecessary renders in the code. To enable this feature, pass `debugSetState: true` to the config and grep JS console logs for `[Onyx-Debug]`.
 
+## Using Redux DevTools extension
+
+It can be useful to check the order of writes to the storage and it's state at a specific point in time.
+
+First, install Redux DevTools through your favorite browser ([Edge](https://microsoftedge.microsoft.com/addons/detail/redux-devtools/nnkgneoiohoecpdiaponcejilbhhikei), [Chrome](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd), [Firefox](https://addons.mozilla.org/en-US/firefox/addon/reduxdevtools/))
+
+Then, you can enable this type of debugging by passing `enableDevTools: true` to `Onyx.init({...})`.
```
Comment (AndrewGable) [inline README.md:334]: I kinda wonder if this should be enabled by default on dev?

Comment (pac-guerreiro) [reply README.md:334]: @AndrewGable I'm not sure if it won't be impactful on performance 😅

**Comment by pac-guerreiro** [comment]
> Hey @pac-guerreiro we have a conflict now, can you please take a look?
> 
> I'll review this tomorrow, got back from OOO today and had to prioritize other issues/reviews.

@PauloGasparSv I just resolved the conflict 😄

**Comment by PauloGasparSv** [comment]
Sorry for the delay, I had time to check the changes here (so far lgtm) but didn't test this yet. Will test it using the extension tomorrow : )

**Comment by pac-guerreiro** [comment]
@PauloGasparSv 

No problem 😄 I just fixed the tests that were broken because of my changes!

File: lib/DevTools.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import {connectViaExtension, extractState} from 'remotedev';
```
Comment (PauloGasparSv) [inline lib/DevTools.js:1]: Hey @pac-guerreiro I'm really curious on how you decided on using `remotedev`! I'm relatively new to  React/ReactNative so would love to know : D

I could only find [this slack post](https://expensify.slack.com/archives/C01GTK53T8Q/p1690385918332959) you sent recently on this.

(btw, this is super cool)

Comment (pac-guerreiro) [reply lib/DevTools.js:1]: @PauloGasparSv 

Sorry for the late response, I'm glad you like this idea 😄 I hope it helps a lot of developers working on the app 😄 

Well, at first I thought I could fork the repo for redux dev tools extension but then I found it was too complex to adapt it as it was for `Onyx`, so after some digging and research I found `remotedev`, which was made by the same guy behind redux dev tools extension, if I'm not mistaken. It basically provides an api to connect to redux dev tools or any other extension that supports `remotedev`, so I gave it a shot and it worked 😄

**Review by PauloGasparSv — APPROVED** [review]
LGTM, this is pretty neat! 

image

All yours @AndrewGable @Gonals

**Comment by Gonals** [comment]
OOO starting tomorrow, so removing myself!

**Comment by PauloGasparSv** [comment]
@pac-guerreiro do you mind fixing the new conflicts we have? 

Bump @AndrewGable for a review : D

**Review by AndrewGable — DISMISSED** [review]

**Comment by PauloGasparSv** [comment]
Hey @pac-guerreiro I think I may have found a bug [in this line](https://github.com/Expensify/react-native-onyx/pull/289/files#diff-146c57cda221f8f5ee5c808eee0c6692b02c4ac2b8818d0104896a04be78e78aR1138) this time:

image

I think the initial `state` can be undefined:

image

Can you please take a look at that?

**Comment by pac-guerreiro** [comment]
@PauloGasparSv 

This issue happens inside `remotedev`'s `extractState` function, in which `JSON.parse` returns this state as `undefined` sometimes 😅 

Honestly, I guess the whole callback passed to DevTools constructor can be discarded since this was meant to update the storage each time we time travel between actions in DevTools but it doesn't work correctly 😅

**Comment by PauloGasparSv** [comment]
> This issue happens inside remotedev's extractState function, in which JSON.parse returns this state as undefined sometimes 😅

Oh I see, I didn't double check but does it return `undefined` even if there is state data there?

If that's the case then it doesn't even make sense to coalesce that value but maybe we can skip that logic that removes the keys in case `newState` is `undefined`

> Honestly, I guess the whole callback passed to DevTools constructor can be discarded since this was meant to update the storage each time we time travel between actions in DevTools but it doesn't work correctly 😅

Really?? So we don't that need that bit of logic?

Sorry for the confusion ahaha I'm trying to understand what we can do next here : )

**Comment by pac-guerreiro** [comment]
@PauloGasparSv 

> Oh I see, I didn't double check but does it return undefined even if there is state data there?

Yeah, I think it has some trouble parsing slash tokens sometimes, maybe because the extension is sending some messages in a format that remotedev is not ready to interpret 😅 

> Really?? So we don't that need that bit of logic?

It could be useful for debugging purposes if a dev wants to go back in time and see the state changes in real-time, but since this logic is broken because either that undefined issue occurs or it does work but going back causes new events to get triggered and shown in the dev tools 😅 

So yeah, if you think it's better, I can drop this logic and we just provide the ability to check state changes in real time 😄

**Comment by PauloGasparSv** [comment]
Bump @AndrewGable : )

**Comment by AndrewGable** [comment]
Yes that sounds good 👍

**Comment by PauloGasparSv** [comment]
Bump @pac-guerreiro :D

**Comment by TomaszFrackowiak** [comment]
@PauloGasparSv - hey, just to let you know that @pac-guerreiro is ooo till 25.08

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv - hey, just to let you know that @pac-guerreiro is ooo till 25.08

Wasn't aware, thanks a lot @TomaszFrackowiak!!!

**Comment by pac-guerreiro** [comment]
Bump @AndrewGable

**Review by AndrewGable — APPROVED** [review]

**Comment by AndrewGable** [comment]
Screenshot 2023-09-05 at 1 49 46 PM

Oops. Looks like your going to have to re-sign your commits.

**Comment by ospfranco** [comment]
Trying to run the tests on the main Expensify repo and this change causes the tests to fail:

CleanShot 2023-09-11 at 10 40 38

Seems the remotedev dependency was added as a dev dependency which is not installed when ran from the main repo.

**Comment by AndrewGable** [comment]
@pac-guerreiro - Can you [address this issue in a PR](https://github.com/Expensify/react-native-onyx/pull/289#issuecomment-1713435305) then re-submit? Looks like this PR was reverted.

**Comment by pac-guerreiro** [comment]
@AndrewGable Hope I'm not late to the party but here's the new PR for this work - https://github.com/Expensify/react-native-onyx/pull/438

I redone everything from scratch since some things got changed in Onyx during this period.

Instead of relying on third party node modules, I decided to implement the logic to connect to the extension by myself after doing some research into `remotedev` repo.

Onyx will connect to Redux DevTools extension by default.

Developers will only be able to use this functionality on web apps.

**Pull request merged**